torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1253,27 +1278,35 @@ def _get_key(self):
         """
         return (len(self.replacements), len(self.divisible))
 
-    def _produce_dyn_sizes(self, ex: torch.Tensor, source: Source) -> List[sympy.Expr]:
+    def _produce_dyn_sizes(self,
+                           ex: torch.Tensor,
+                           source: Source,
+                           dims: Optional[List[DIM_DYNAMISM_STATE]]) -> List[sympy.Expr]:
         from torch._dynamo.source import TensorPropertySource, TensorProperty
         size = []
         for i, val in enumerate(ex.size()):
-            is_dynamic = _is_dim_dynamic(ex, i)
-            if _should_allocate(is_dynamic, self.assume_static_by_default):
+            dim_state = dims[i] if dims else DIM_DYNAMISM_STATE.DUCK
+            is_static = dim_state == DIM_DYNAMISM_STATE.STATIC
+            if not is_static:
                 size.append(self.create_symbol(
-                    val, TensorPropertySource(source, TensorProperty.SIZE, i), is_dynamic
+                    val, TensorPropertySource(source, TensorProperty.SIZE, i), dim_state
                 ))
             else:
                 size.append(sympy.Integer(val))
         return size
 
-    def create_symbolic_sizes_strides_storage_offset(self, ex: torch.Tensor, source: Source):
+    def create_symbolic_sizes_strides_storage_offset(self,
+                                                     ex: torch.Tensor,
+                                                     source: Source,
+                                                     dynamic_dims: Optional[List[DIM_DYNAMISM_STATE]],
+                                                     dynamic_dims_range: Dict[int, "MinMaxConstraint"]):
```

Comment:
Oh, you implemented up-front? It's weird that dynamic_dims is a list but dynamic_dims_range is a dict; they are both keyed on dims.

Reply:
dynamic_dims is 1:1 with the size, every dim has a state.

dynamic_dims_range is only for user marked

Reply:
I mean, I could just as easily say that dynamic_dims is either DYNAMIC or DUCK, and if it's missing then it's STATIC. Would you prefer it if they were all dicts? I mostly just want these to be uniform. (Similarly, Optional[List is a stick in my craw, can you really pass None here?!

Reply:
> I mean, I could just as easily say that dynamic_dims is either DYNAMIC or DUCK, and if it's missing then it's STATIC.

True, I do not have a strong preference between that state and this.

> Would you prefer it if they were all dicts? I mostly just want these to be uniform.

If I don't have a strong pref, and you do, yours wins. Let's make them both dicts, and if an index is missing we assume static.

> (Similarly, Optional[List is a stick in my craw, can you really pass None here?!

You asked that above ;) And yes, we can, but we can make this required and remove all "default" notions here...